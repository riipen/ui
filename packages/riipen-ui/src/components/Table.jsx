import clsx from "clsx";
import PropTypes from "prop-types";
import React from "react";

import ThemeContext from "../styles/ThemeContext";
import withClasses from "../utils/withClasses";
import withForwardedRef from "../utils/withForwardedRef";

class Table extends React.Component {
  static propTypes = {
    /**
     * What background to render for the table.
     * Defaults to grey100.
     */
    backgroundColor: PropTypes.oneOf([
      "transparent",
      "grey50",
      "grey100",
      "grey200",
      "grey300",
      "grey400"
    ]),

    /**
     * Children to render in the table element.
     */
    children: PropTypes.node,

    /**
     * Classes to apply to the root element.
     */
    classes: PropTypes.arrayOf(PropTypes.string),

    /**
     * A reference to the table DOM element.
     */
    forwardedRef: PropTypes.element,

    /**
     * Whether to give all columns in the table equal width.
     */
    layout: PropTypes.oneOf(["auto", "fixed"])
  };

  static defaultProps = {
    backgroundColor: "grey100",
    layout: "auto"
  };

  static contextType = ThemeContext;

  render() {
    const {
      children,
      classes,
      layout,
      backgroundColor,
      forwardedRef
    } = this.props;

    const theme = this.context;

    const tableBackground = {
      transparent: "transparent",
      grey50: theme.palette.grey[50],
      grey100: theme.palette.grey[100],
      grey200: theme.palette.grey[200],
      grey300: theme.palette.grey[300],
      grey400: theme.palette.grey[400]
    };

    return (
      <React.Fragment>
        <table ref={forwardedRef} className={clsx(classes)} border="0">
          {children}
        </table>
        <style jsx>{`
          table {
            background-color: ${tableBackground[backgroundColor]};
            border-radius: 4px;
            border-collapse: collapse;
            table-layout: ${layout};
            width: 100%;
          }
        `}</style>
      </React.Fragment>
    );
  }
}

export default withForwardedRef(withClasses(Table));
