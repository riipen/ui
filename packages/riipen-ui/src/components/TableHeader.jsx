import clsx from "clsx";
import PropTypes from "prop-types";
import React from "react";

import withClasses from "../utils/withClasses";

class TableHeader extends React.Component {
  static propTypes = {
    /**
     * Children to render in the thead element
     */
    children: PropTypes.node,

    /**
     * Classes to apply to the root element
     */
    classes: PropTypes.arrayOf(PropTypes.string)
  };

  render() {
    const { children, classes, ...other } = this.props;
    return (
      <thead className={clsx(classes)} {...other}>
        {children}
      </thead>
    );
  }
}

export default withClasses(TableHeader);
