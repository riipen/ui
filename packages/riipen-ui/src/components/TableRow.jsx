import clsx from "clsx";
import PropTypes from "prop-types";
import React from "react";

import ThemeContext from "../styles/ThemeContext";
import withClasses from "../utils/withClasses";

class TableRow extends React.Component {
  static propTypes = {
    /**
     * Children to render in the tr element
     */
    children: PropTypes.node,

    /**
     * Classes to apply to the root element
     */
    classes: PropTypes.arrayOf(PropTypes.string),

    /**
     * Whether to highlight the row on hover
     */
    hover: PropTypes.bool,

    /**
     * Whether to render the border for the row
     */
    border: PropTypes.bool
  };

  static defaultProps = {
    border: true
  };

  static contextType = ThemeContext;

  render() {
    const { children, classes, hover, border } = this.props;

    const theme = this.context;

    return (
      <React.Fragment>
        <tr className={clsx(classes)}>{children}</tr>
        <style jsx>{`
          tr {
            border-bottom: 1px solid
              ${border &&
                (theme.palette.divider
                  ? theme.palette.divider
                  : `transparent`)};
            border-left: 3px solid transparent;
          }

          tr:hover {
            ${hover ? `background-color: ${theme.palette.common.white};` : ""}
            ${hover
              ? `border-left: 3px solid ${theme.palette.secondary.main};`
              : ""}
          }
        `}</style>
      </React.Fragment>
    );
  }
}

export default withClasses()(TableRow);
