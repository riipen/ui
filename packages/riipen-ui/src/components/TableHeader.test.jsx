import { mount } from "enzyme";
import toJson from "enzyme-to-json";
import React from "react";

import Table from "./Table.jsx";
import TableHeader from "./TableHeader.jsx";
import TableRow from "./TableRow.jsx";

describe("<TableHeader>", () => {
  it("renders without errors", () => {
    let error;

    try {
      mount(
        <Table>
          <TableHeader />
        </Table>
      );
    } catch (theError) {
      error = theError;
    }

    expect(error).toEqual(undefined);
  });

  it("renders correct snapshot", () => {
    const wrapper = mount(
      <Table>
        <TableHeader />
      </Table>
    );

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  describe("children prop", () => {
    it("renders given non null children", () => {
      const child = <TableRow />;

      const wrapper = mount(
        <Table>
          <TableHeader>
            {null}
            {child}
          </TableHeader>
        </Table>
      );

      expect(
        wrapper
          .find("thead")
          .childAt(0)
          .matchesElement(child)
      ).toEqual(true);
    });
  });

  describe("classes prop", () => {
    it("applies classes to the root node", () => {
      const classes = ["classOne"];

      const wrapper = mount(
        <Table>
          <TableHeader classes={classes} />
        </Table>
      );

      expect(wrapper.find("thead").hasClass(classes[0])).toEqual(true);
    });

    it("appends higher order values to classes prop with withClasses decorator", () => {
      const classes = ["riipen", "riipen-tableheader"];

      const wrapper = mount(
        <Table>
          <TableHeader />
        </Table>
      );

      expect(
        wrapper
          .find("TableHeader")
          .props()
          .classes.sort()
      ).toEqual(classes.sort());
    });
  });
});
