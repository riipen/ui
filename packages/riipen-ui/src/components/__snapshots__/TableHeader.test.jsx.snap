// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TableHeader> renders correct snapshot 1`] = `
<ForwardRef(withForwardedRef(withClasses(Table)))>
  <withClasses(Table)
    forwardedRef={null}
  >
    <Table
      backgroundColor="grey50"
      classes={
        Array [
          "riipen",
          "riipen-table",
        ]
      }
      forwardedRef={null}
      layout="auto"
    >
      <table
        border="0"
        className="jsx-1059775881 riipen riipen-table"
      >
        <withClasses(TableHeader)>
          <TableHeader
            classes={
              Array [
                "riipen",
                "riipen-tableheader",
              ]
            }
          >
            <thead
              className="riipen riipen-tableheader"
            />
          </TableHeader>
        </withClasses(TableHeader)>
      </table>
      <JSXStyle
        dynamic={
          Array [
            "#fafafa",
            "auto",
          ]
        }
        id="4123435624"
      />
    </Table>
  </withClasses(Table)>
</ForwardRef(withForwardedRef(withClasses(Table)))>
`;
